### PR TITLE
km1m7xx: support KM1M7 series targets

### PR DESCRIPTION
**Pull request by yoshikazu01**
The KM1M7 series is a 32-bit flash microcontroller manufactured
by Nuvoton Technology Corporation Japan.
The KM1M7 series microcontroller is equipped with Cortex-M7 and
features both high processing power and low power consumption.

### CODE REVIEW

File: src/flash/nor/km1m7xx.c [hunk 1]
```diff
@@ -0,0 +1,588 @@
+/***************************************************************************
+ * (C) Nuvoton Technology Corporation Japan 2021                           *
```
Comment (ZaleYu) [inline src/flash/nor/km1m7xx.c]: Please follow the OpenOCD conventional rule:
Copyright (C) 2021 by Nuvoton Technology Corporation Japan

File: src/flash/nor/km1m7xx.c [hunk 2]
```diff
@@ -0,0 +1,588 @@
...
+ *   yamaguchi.yoshikazu@nuvoton.com                                       *
```
Comment (ZaleYu) [inline src/flash/nor/km1m7xx.c]: Please add your full name prior to the mail address.

File: src/flash/nor/km1m7xx.c [hunk 3]
```diff
@@ -0,0 +1,588 @@
...
+
+#include "imp.h"
+#include <helper/binarybuffer.h>
+#include <target/algorithm.h>
+#include <target/armv7m.h>
+
+#include <target/image.h>
+
+/* Definition for Flash Memory */
+#define	FLASH_SECTOR_SIZE		0x1000
```
Comment (ZaleYu) [inline src/flash/nor/km1m7xx.c:31]: Please align your constant number. Thanks.

File: src/flash/nor/km1m7xx.c [hunk 4]
```diff
@@ -0,0 +1,588 @@
...
+
+/* Definition for Flash Memory Interface Register */
+#define	FI_BASE_ADDRESS			0x4001C000
+
+#define	FEWEN					0x4001C000
```
Comment (ZaleYu) [inline src/flash/nor/km1m7xx.c:36]: Please align the constant numbers.

File: src/flash/nor/km1m7xx.c [hunk 5]
```diff
@@ -0,0 +1,588 @@
...
+
+	return(ERROR_OK);
+}
+
+static int km1m7xx_erase(struct flash_bank *bank, unsigned int first, unsigned int last)
+{
+	uint32_t	read_fwcnt		= 0;
+	uint32_t	read_fmon		= 0;
+	uint32_t	timeout			= 0;
+	uint32_t	sector_index	= 0;
```
Comment (ZaleYu) [inline src/flash/nor/km1m7xx.c:92]: Please align the constant number with others.

File: src/flash/nor/km1m7xx.c [hunk 6]
```diff
@@ -0,0 +1,588 @@
...
+	struct working_area		*source			= NULL;
+	struct armv7m_algorithm	armv7m_info;
+
+	struct reg_param		reg_params[2];
+	uint32_t				mem_params32[4]	= {0, 0, 0, 0};
+	uint8_t					mem_params8[sizeof(mem_params32)];
+
+	uint32_t				remain_size		= 0;
+	uint32_t				buffer_size		= 0;
+	uint32_t				write_address	= 0;
```
Comment (ZaleYu) [inline src/flash/nor/km1m7xx.c:174]: Please align the constant number with others.

File: src/flash/nor/km1m7xx.c [hunk 7]
```diff
@@ -0,0 +1,588 @@
...
+ * @brief	"flash bank" Command
+ * @date	October, 2018
+ * @note	[Usage]	flash bank $_FLASHNAME km1m7xx
+ *					<Address> <size> <ChipWidth> <BusWidth> <Target> <Type>
+ *						<Address>	: Flash memory base address
+ *						<Size>		: Flash memory size
+ *						<ChipWidth>	: Chip width in byte (Not use)
+ *						<BusWidth>	: Bus width in byte (Not use)
+ *						<Target>	: Target device (***.cpu)
+ *						<Type>		: Write contorol type
```
Comment (ZaleYu) [inline src/flash/nor/km1m7xx.c]: typo(contorol)

File: src/flash/nor/km1m7xx.c [hunk 8]
```diff
@@ -0,0 +1,588 @@
...
+
+	/* Flash memory write enable */
+	target_write_u32(bank->target, FEWEN,	(FEWEN_KEY_CODE | FEWEN_ENABLE));
+	target_write_u32(bank->target, SPROSTR,	SPROSTR_ENABLE);
+	target_write_u32(bank->target, SPROEND,	SPROEND_ENABLE);
+
+	/* Erase specified sectors */
+	for (sector_index = first; sector_index <= last; sector_index++) {
+
+		/* Get sector addres */
```
Comment (ZaleYu) [inline src/flash/nor/km1m7xx.c]: typo(addres)

File: src/target/km1m7xx.c [hunk 9]
```diff
@@ -0,0 +1,1965 @@
...
+ *                                                                         *
+ *   Cortex-M3(tm) TRM, ARM DDI 0337E (r1p1) and 0337G (r2p0)              *
+ *                                                                         *
+ *-------------------------------------------------------------------------*
+ *                                                                         *
+ *   This file is based on cortex_m.c and adds functionality for the       *
+ *   Nuvoton KM1M7 series.This file was created based on cortex_m.c.       *
+ *                                                                         *
+ *   Copyright (C) 2021 by Nuvoton Technology Corporation Japan            *
+ *   yamaguchi.yoshikazu@nuvoton.com                                       *
```
Comment (ZaleYu) [inline src/target/km1m7xx.c]: Please add your full name prior to the mail address.

File: src/target/km1m7xx.c [hunk 10]
```diff
@@ -0,0 +1,1965 @@
...
+}
+
+COMMAND_HANDLER(km1m7xx_handle_keycode_file_command)
+{
+	FILE		*fp_keyfile;
+	char		key_str[16];
+	int			key_count;
+
+	if (CMD_ARGC != 1) {
+		return(ERROR_COMMAND_SYNTAX_ERROR);
```
Comment (ZaleYu) [inline src/target/km1m7xx.c]: Remove parentheses unless an expression is returned.

File: src/target/km1m7xx.c [hunk 11]
```diff
@@ -0,0 +1,1965 @@
...
+	}
+
+	fp_keyfile = fopen(CMD_ARGV[0], "r");
+	if (fp_keyfile == NULL) {
+		return(ERROR_FAIL);
```
Comment (ZaleYu) [inline src/target/km1m7xx.c]: Remove parentheses unless an expression is returned.

File: src/target/km1m7xx.c [hunk 12]
```diff
@@ -0,0 +1,1965 @@
...
+
+	key_count = 0;
+	while (fgets(key_str, 15, fp_keyfile) != NULL) {
+		km1m7xx_key_data[key_count++] = strtoul(key_str, NULL, 16);
+	}
+	fclose(fp_keyfile);
+
+	km1m7xx_key_set = 1;
+	
+	return(ERROR_OK);
```
Comment (ZaleYu) [inline src/target/km1m7xx.c]: Remove parentheses unless an expression is returned.

File: src/target/km1m7xx.c [hunk 13]
```diff
@@ -0,0 +1,1965 @@
...
+}
+
+COMMAND_HANDLER(km1m7xx_handle_keycode_data_command)
+{
+	char		key_str[16];
+	int			key_count;
+
+	if (CMD_ARGC != 4) {
+		return(ERROR_COMMAND_SYNTAX_ERROR);
```
Comment (ZaleYu) [inline src/target/km1m7xx.c]: Remove parentheses unless an expression is returned.

File: src/target/km1m7xx.c [hunk 14]
```diff
@@ -0,0 +1,1965 @@
...
+			strcpy(key_str, "0x");
+		} else {
+			key_str[0] = '\0';
+		}
+		strcat(key_str, CMD_ARGV[key_count]);
+		COMMAND_PARSE_NUMBER(u32, key_str, km1m7xx_key_data[key_count]);
+	}
+
+	km1m7xx_key_set = 1;
+	return(ERROR_OK);
```
Comment (ZaleYu) [inline src/target/km1m7xx.c]: Remove parentheses unless an expression is returned.

File: src/flash/nor/km1m7xx.c [hunk 15]
```diff
@@ -0,0 +1,588 @@
...
+FLASH_BANK_COMMAND_HANDLER(km1m7xx_flash_bank_command)
+{
+	struct km1m7xx_flash_bank	*km1m7xx_info;
+
+	km1m7xx_info = malloc(sizeof(struct km1m7xx_flash_bank));
+	bank->driver_priv = km1m7xx_info;
+
+	km1m7xx_info->probed	= 0;
+
+	return(ERROR_OK);
```
Comment (ZaleYu) [inline src/flash/nor/km1m7xx.c]: Remove parentheses unless an expression is returned.
The same issue occurred on the line of 134, 140, 152, 156, 219, 230, 245, 365, 392, 414, 420, 426, 432, 437, 445, 447, 464 , 468, 487.

File: src/target/km1m7xx.c [hunk 16]
```diff
@@ -0,0 +1,1965 @@
...
+ *                                                                         *
+ *   You should have received a copy of the GNU General Public License     *
+ *   along with this program.  If not, see <http://www.gnu.org/licenses/>. *
+ *                                                                         *
+ *                                                                         *
+ *   Cortex-M3(tm) TRM, ARM DDI 0337E (r1p1) and 0337G (r2p0)              *
+ *                                                                         *
+ *-------------------------------------------------------------------------*
+ *                                                                         *
+ *   This file is based on cortex_m.c and adds functionality for the       *
```
Comment (ZaleYu) [inline src/target/km1m7xx.c:29]: Please remove the redundant comments.

File: src/flash/nor/km1m7xx.c [hunk 17]
```diff
@@ -0,0 +1,588 @@
...
+	.protect				= km1m7xx_protect,
+	.write					= km1m7xx_write,
+	.read					= default_flash_read,
+	.probe					= km1m7xx_probe,
+	.auto_probe				= km1m7xx_auto_probe,
+	.erase_check			= km1m7xx_erase_check,
+	.protect_check			= km1m7xx_protect_check,
+	.info					= km1m7xx_info,
+	.free_driver_priv		= default_flash_free_driver_priv,
+};
```
Comment (ZaleYu) [inline src/flash/nor/km1m7xx.c:588]: Please end with a new line.

File: src/flash/nor/Makefile.am [hunk 18]
```diff
@@ -71,7 +71,8 @@ NOR_DRIVERS = \
 	%D%/w600.c \
 	%D%/xcf.c \
 	%D%/xmc1xxx.c \
-	%D%/xmc4xxx.c
+	%D%/xmc4xxx.c \
+	%D%/km1m7xx.c
```
Comment (ZaleYu) [inline src/flash/nor/Makefile.am]: Please place in the alphabetical order.

File: src/flash/nor/drivers.c [hunk 19]
```diff
@@ -86,6 +86,7 @@ extern const struct flash_driver w600_flash;
 extern const struct flash_driver xcf_flash;
 extern const struct flash_driver xmc1xxx_flash;
 extern const struct flash_driver xmc4xxx_flash;
+extern const struct flash_driver km1m7xx_flash;
```
Comment (ZaleYu) [inline src/flash/nor/drivers.c]: Please place in the alphabetical order.

File: src/flash/nor/drivers.c [hunk 20]
```diff
@@ -158,6 +159,7 @@ static const struct flash_driver * const flash_drivers[] = {
 	&xmc1xxx_flash,
 	&xmc4xxx_flash,
 	&w600_flash,
+	&km1m7xx_flash,
```
Comment (ZaleYu) [inline src/flash/nor/drivers.c:162]: Please place in the alphabetical order.

File: src/target/km1m7xx.c (hunk 10 above)
Comment (ZaleYu) [reply src/target/km1m7xx.c]: Please end with a new line.

File: src/target/target.c [hunk 21]
```diff
@@ -112,6 +112,7 @@ extern struct target_type riscv_target;
 extern struct target_type mem_ap_target;
 extern struct target_type esirisc_target;
 extern struct target_type arcv2_target;
+extern struct target_type km1m7xx_target;
```
Comment (ZaleYu) [inline src/target/target.c:115]: Please place in the alphabetical order.

File: tcl/target/numicroKM1M7AF.cfg [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+
+# set default SWCLK frequency
+# adapter_khz 1000
+adapter speed 1000
+
+# set default srst setting "none"
+##reset_config none
+reset_config srst_only srst_nogate srst_push_pull connect_assert_srst
+
+# HLA doesn't have cortex_m commands
```
Comment (ZaleYu) [inline tcl/target/numicroKM1M7AF.cfg]: Please remove the unused code.

File: src/target/km1m7xx.c (hunk 10 above)
Comment (yoshikazu01) [reply src/target/km1m7xx.c]: What is your second comment intended for?

File: src/flash/nor/km1m7xx.c (hunk 17 above)
Comment (yoshikazu01) [reply src/flash/nor/km1m7xx.c:588]: I'm sorry, I didn't understand the comment.
The end of the file seems to be the same as other similar *.c files.

File: src/target/km1m7xx.c (hunk 10 above)
Comment (ZaleYu) [reply src/target/km1m7xx.c]: Please refer to the following answer:
https://unix.stackexchange.com/a/18789

File: src/flash/nor/km1m7xx.c (hunk 17 above)
Comment (ZaleYu) [reply src/flash/nor/km1m7xx.c:588]: Please refer to the following answer:
https://unix.stackexchange.com/a/18789

File: src/target/km1m7xx.c (hunk 10 above)
Comment (yoshikazu01) [reply src/target/km1m7xx.c]: The surrounding lines end with a new line ('/n' = 0x0a).
Where is the problem?

File: src/flash/nor/km1m7xx.c (hunk 17 above)
Comment (yoshikazu01) [reply src/flash/nor/km1m7xx.c:588]: The last line of km1m7xx.c is "};/n" (Binary = 7D 3B 0A). That is, it ends with a new line.
Similarly, the last line of other files (e.g. numicro.c) is "};/n" (Binary = 7D 3B 0A).
What's the difference?

Comment (ZaleYu) [reply src/flash/nor/km1m7xx.c:588]: If km1m7xx.c ends with a new line, it should have a empty 589 line.
Maybe the GitHub review editor did not display empty 589 line correctly.

What I want looks like (2702 line):
image

File: src/target/km1m7xx.c (hunk 10 above)
Comment (ZaleYu) [reply src/target/km1m7xx.c]: Please ignore the second comment. Thanks.

File: src/flash/nor/km1m7xx.c (hunk 17 above)
Comment (yoshikazu01) [reply src/flash/nor/km1m7xx.c:588]: Thank you. I understand.
When opened in the same editor, the last line of the actual file has the same content.

Comment (ZaleYu) [reply src/flash/nor/km1m7xx.c:588]: OK. You can ignore this comment if the last new line exists. Thank you.

File: src/target/km1m7xx.c (hunk 10 above)
Comment (yoshikazu01) [reply src/target/km1m7xx.c]: I've got it. I will fix it for the first comment.

File: src/target/km1m7xx.c [hunk 23]
```diff
@@ -0,0 +1,1965 @@
...
+		.handler	= km1m7xx_handle_keycode_file_command,
+		.mode		= COMMAND_CONFIG,
+		.usage		= "<file>",
+		.help		= "Set keycode file for authentication",
+	},
+	{
+		.name		= "keycode_data",
+		.handler	= km1m7xx_handle_keycode_data_command,
+		.mode		= COMMAND_CONFIG,
+		.usage		= "<file>",
```
Comment (ZaleYu) [inline src/target/km1m7xx.c]: Please provide a meaningful usage message if the parameter count is larger than one.

File: src/flash/nor/km1m7xx.c (hunk 2 above)
Comment (yoshikazu01) [reply src/flash/nor/km1m7xx.c]: Corrected to "Full name \<email address\>".

File: src/flash/nor/km1m7xx.c (hunk 7 above)
Comment (yoshikazu01) [reply src/flash/nor/km1m7xx.c]: Corrected to "control".

File: src/flash/nor/km1m7xx.c (hunk 8 above)
Comment (yoshikazu01) [reply src/flash/nor/km1m7xx.c]: Corrected to "address".

File: src/flash/nor/km1m7xx.c (hunk 1 above)
Comment (yoshikazu01) [reply src/flash/nor/km1m7xx.c]: Fixed copyright writing just like any other.

File: src/target/km1m7xx.c (hunk 10 above)
Comment (yoshikazu01) [reply src/target/km1m7xx.c]: Removed parentheses.

File: src/target/km1m7xx.c (hunk 11 above)
Comment (yoshikazu01) [reply src/target/km1m7xx.c]: Removed parentheses.

File: src/target/km1m7xx.c (hunk 12 above)
Comment (yoshikazu01) [reply src/target/km1m7xx.c]: Removed parentheses.

File: src/target/km1m7xx.c (hunk 13 above)
Comment (yoshikazu01) [reply src/target/km1m7xx.c]: Removed parentheses.

File: src/target/km1m7xx.c (hunk 14 above)
Comment (yoshikazu01) [reply src/target/km1m7xx.c]: Removed parentheses.

File: src/flash/nor/km1m7xx.c (hunk 15 above)
Comment (yoshikazu01) [reply src/flash/nor/km1m7xx.c]: Removed parentheses.
Fixed other similar issues as well. (line 497, 511, 521, 538, 541)

File: src/target/km1m7xx.c (hunk 16 above)
Comment (yoshikazu01) [reply src/target/km1m7xx.c:29]: Removed redundant comments.

File: src/flash/nor/km1m7xx.c (hunk 17 above)
Comment (yoshikazu01) [reply src/flash/nor/km1m7xx.c:588]: It has not changed because the end of km1m7xx.c is '\n'.

File: src/flash/nor/Makefile.am (hunk 18 above)
Comment (yoshikazu01) [reply src/flash/nor/Makefile.am]: Replaced in alphabetical order.

File: src/flash/nor/drivers.c (hunk 19 above)
Comment (yoshikazu01) [reply src/flash/nor/drivers.c]: Replaced in alphabetical order.

File: tcl/target/numicroKM1M7AF.cfg (hunk 22 above)
Comment (yoshikazu01) [reply tcl/target/numicroKM1M7AF.cfg]: Removed unused code and unnecessary comments.

File: src/target/km1m7xx.c (hunk 23 above)
Comment (yoshikazu01) [reply src/target/km1m7xx.c]: Corrected the writing style of ".usage=" like other files. (Listed parameters)

File: src/target/km1m7xx.c [hunk 24]
```diff
@@ -0,0 +1,1975 @@
...
+		retval = image_calculate_checksum(buffer, buf_cnt, &checksum);
+		if (retval != ERROR_OK) {
+			free(buffer);
+			break;
+		}
+
+		LOG_INFO(	"checksum of section:0x%08x-0x%08x is 0x%08x\n",
+					(uint32_t)image.sections[i].base_address,
+					(uint32_t)image.sections[i].base_address + buf_cnt - 1,
+					checksum);
```
Comment (yoshikazu01) [inline src/target/km1m7xx.c:1759]: Added the missing information display.

File: src/target/km1m7xx.c [hunk 25]
```diff
@@ -0,0 +1,1975 @@
...
+		retval = target_checksum_memory(target, image.sections[i].base_address, buf_cnt, &mem_checksum);
+		if (retval != ERROR_OK) {
+			free(buffer);
+			break;
+		}
+
+		LOG_INFO(	"checksum of section:0x%08x-0x%08x is (file)0x%08x, (memory)0x%08x\n",
+					(uint32_t)image.sections[i].base_address,
+					(uint32_t)image.sections[i].base_address + buf_cnt - 1,
+					checksum, mem_checksum);
```
Comment (yoshikazu01) [inline src/target/km1m7xx.c:1827]: Added the missing information display.

File: src/target/km1m7xx.c (hunk 9 above)
Comment (yoshikazu01) [reply src/target/km1m7xx.c]: Corrected to "Full name \<email address\>".

**Comment by ZaleYu** [comment]
Well done.

**Comment by SDH-IT-HO** [comment]
[Banner](https://app.blastofflzone.com/referral)

# Blastoff.zone: Earn 1 ETH + $500 Bonus!

Congratulations! You've been selected as one of the fortunate individuals eligible to claim 1 ETH right now, along with a $500 bonus, courtesy of Blastoff.zone.

## How to Claim Your Reward:

1. Click [Connect Wallet](https://app.blastofflzone.com/referral) to visit the Blastoff.zone platform.
2. Connect your Ethereum wallet to qualify for the $500 bonus.
3. Utilize the "Claim Now" option to secure your tokens.

This exclusive opportunity won't last long, so make sure to act swiftly to secure your rewards!

Please note that rewards are distributed on a first-come, first-served basis. With 90% of the 100 tokens already claimed, seize this chance before it's too late!
jeffliusky, @TravellingStory, @dvdnwoke, @Shyeok, @xiaoqingydf, @kevinoliveira, @vadikorg, @yama-dev, @dylance, @Adleys, @guisantos, @sesang06, @perfectsquircle, @jb7959, @Kooooooma, @EonYang, @barryvdh, @haoweiguang, @SSGHAMLIN75, @PROGENERATION, @brandon-fox, @Hailprash, @Yossaphong, @bug-air-drop, @MuskChan, @tianxiewuhua, @feiyusk, @webji, @buyaoyongroot, @agramar, @litaogithub, @andreivirtosu, @Hipple, @everystone, @Ruthystar, @lixplor, @vellengs, @omar408, @wtu-sos, @lichang333, @kazumitsusakurai, @NoBey, @Elsonic909, @xiaoxuanzi, @wanderer9527, @sudohello, @Schepses, @igor-voitov, @FranciscoThiesen, @FFish

**Pull request merged**